File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.js
/**
 * @param {number[][]} matrix
 * @param {number} target
 * @return {boolean}
 */
var searchMatrix = function(matrix, target) {
  
    let m = matrix.length
    let n = matrix[0].length
    
    let left = 0
    let right = m * n - 1
    
    while (left <= right){
        let mid = Math.floor((left + right) / 2)
        
        let mid_row = Math.floor(mid / n)
        let mid_col = mid % n
        
        if (matrix[mid_row][mid_col] == target){
            return true
        }
        
        if (matrix[mid_row][mid_col] < target){
            left = mid + 1
        }
        
        if (matrix[mid_row][mid_col] > target){
            right = mid - 1
        }
    }
    
    return false
};

File: 0074-search-a-2d-matrix/0074-search-a-2d-matrix.py
class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
            
        def bsearch(arr, target):
            n = len(arr)
        
            left, right = 0, n - 1

            while left <= right:
                mid = (left + right) // 2
                mid_value = arr[mid]

                if mid_value == target:
                    return True

                if mid_value > target:
                    right = mid - 1

                if mid_value < target:
                    left = mid + 1

            return False
            
        for arr in matrix:
            if bsearch(arr, target):
                return True
        
        return False
        